pkg/karmadactl/join.go

```diff
@@ -187,66 +187,45 @@ func JoinCluster(controlPlaneRestConfig, clusterConfig *rest.Config, opts Comman
 		return nil
 	}
 
-	// create secret in control plane
-	secret := &corev1.Secret{
-		ObjectMeta: metav1.ObjectMeta{
-			Namespace: opts.ClusterNamespace,
-			Name:      opts.ClusterName,
-		},
-		Data: map[string][]byte{
-			clusterv1alpha1.SecretCADataKey: clusterSecret.Data["ca.crt"],
-			clusterv1alpha1.SecretTokenKey:  clusterSecret.Data[clusterv1alpha1.SecretTokenKey],
-		},
-	}
-	secret, err = util.CreateSecret(controlPlaneKubeClient, secret)
-	if err != nil {
-		return fmt.Errorf("failed to create secret in control plane. error: %v", err)
-	}
-
-	cluster, err := generateClusterInControllerPlane(controlPlaneRestConfig, clusterConfig, opts, *secret)
+	err = registerClusterInControllerPlane(opts, controlPlaneRestConfig, clusterConfig, controlPlaneKubeClient, clusterSecret, impersonationSecret)
 	if err != nil {
 		return err
 	}
 
-	patchSecretBody := &corev1.Secret{
-		ObjectMeta: metav1.ObjectMeta{
-			OwnerReferences: []metav1.OwnerReference{
-				*metav1.NewControllerRef(cluster, clusterResourceKind),
-			},
-		},
-	}
-	err = util.PatchSecret(controlPlaneKubeClient, secret.Namespace, secret.Name, types.MergePatchType, patchSecretBody)
-	if err != nil {
-		return fmt.Errorf("failed to patch secret %s/%s, error: %v", secret.Namespace, secret.Name, err)
-	}
-
 	fmt.Printf("cluster(%s) is joined successfully\n", opts.ClusterName)
-
 	return nil
 }
 
-func generateSecretInMemberCluster(clusterKubeClient kubeclient.Interface, clusterNamespace, clusterName string, dryRun bool) (*corev1.Secret, error) {
+func obtainCredentialsFromMemberCluster(clusterKubeClient kubeclient.Interface, clusterNamespace, clusterName string, dryRun bool) (*corev1.Secret, *corev1.Secret, error) {
 	var err error
 
 	// ensure namespace where the karmada control plane credential be stored exists in cluster.
 	if _, err = ensureNamespaceExist(clusterKubeClient, clusterNamespace, dryRun); err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	// create a ServiceAccount in cluster.
 	serviceAccountObj := &corev1.ServiceAccount{}
 	serviceAccountObj.Namespace = clusterNamespace
 	serviceAccountObj.Name = names.GenerateServiceAccountName(clusterName)
 	if serviceAccountObj, err = ensureServiceAccountExist(clusterKubeClient, serviceAccountObj, dryRun); err != nil {
-		return nil, err
+		return nil, nil, err
+	}
+
+	// create a ServiceAccount for impersonation in cluster.
+	impersonationSA := &corev1.ServiceAccount{}
+	impersonationSA.Namespace = clusterNamespace
+	impersonationSA.Name = names.GenerateServiceAccountName("impersonator")
+	if impersonationSA, err = ensureServiceAccountExist(clusterKubeClient, impersonationSA, dryRun); err != nil {
+		return nil, nil, err
```

Comment:
What we need is to create a `SA` on the member cluster. I think we can abstract another function instead of coupling it with the current one.